olive/hardware/accelerator.py

```diff
@@ -92,89 +92,192 @@ def filter_execution_providers(execution_providers, available_providers):
         return [ep for ep in available_providers if ep in execution_providers]
 
     @staticmethod
-    def infer_accelerators_from_execution_provider(execution_provider: List[str]):
+    def infer_accelerators_from_execution_providers(execution_providers: List[str]):
         """Infer the device from the execution provider name.
 
         If all the execution provider is uniquely mapped to a device, return the device list.
         Otherwise, return None.
+        Please note that the CPUExecutionProvider is skipped for device infer. And only other ORT EPs are considered.
         For example:
             execution_provider = ["CPUExecutionProvider", "CUDAExecutionProvider"]
-            return None (CPUExecutionProvider is mapped to CPU and GPU, Olive cannot infer the device)
+            return ["gpu"]
             execution_provider = ["CUDAExecutionProvider", "TensorrtExecutionProvider"]
             return ["gpu"]
         """
-        if not execution_provider:
+        if not execution_providers:
             return None
 
-        is_unique_inferring = True
-        accelerators = []
-        for idx, ep in enumerate(execution_provider):
-            accelerators.append([])
+        ep_to_accelerator = {}
+        for ep in execution_providers:
+            if ep == "CPUExecutionProvider":
+                # cannot infer device for CPUExecutionProvider since all ORT EP supports CPU
```

Comment:
As a special case, If CPUExecutionProvider is explicitly listed alone (without any other EPs) then it should be sufficient to infer CPU.

Reply:
Mike told me he has added this as a special case already. It is here https://github.com/microsoft/Olive/blob/9c7dc9f7303b5527857ca8bd183a59beb601135e/olive/hardware/accelerator.py#L203

Reply:
Yeah, @jambayk is right. The logic is handled in the caller, If the EPs(after removing AzureExecutionProvider) is only one "CPUExecutionProvider", the inferred device will be cpu.